x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+
+    private final String fieldName;
+    private final String modelText;
+    private final String modelId;
+    private SetOnce<SlimResults> weightedTokensSupplier;
+    private QueryBuilder termsQuery;
+
+    public TextExpansionQueryBuilder(String fieldName, String modelText, String modelId) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (modelText == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + MODEL_TEXT.getPreferredName() + " value");
+        }
+
+        this.fieldName = fieldName;
+        this.modelText = modelText;
+        this.modelId = modelId == null ? DEFAULT_MODEL_ID : modelId;
+    }
+
+    public TextExpansionQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.modelText = in.readString();
+        this.modelId = in.readString();
+        this.termsQuery = in.readOptionalNamedWriteable(QueryBuilder.class);
+    }
+
+    private TextExpansionQueryBuilder(TextExpansionQueryBuilder other, SetOnce<SlimResults> weightedTokensSupplier) {
+        this.fieldName = other.fieldName;
+        this.modelText = other.modelText;
+        this.modelId = other.modelId;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.weightedTokensSupplier = weightedTokensSupplier;
+    }
+
+    private TextExpansionQueryBuilder(TextExpansionQueryBuilder other, QueryBuilder rewrittenTermsQuery) {
+        this.fieldName = other.fieldName;
+        this.modelText = other.modelText;
+        this.modelId = other.modelId;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.termsQuery = rewrittenTermsQuery;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_8_0;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeString(modelText);
+        out.writeString(modelId);
+        out.writeOptionalNamedWriteable(termsQuery);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.field(MODEL_TEXT.getPreferredName(), modelText);
+        builder.field(MODEL_ID.getPreferredName(), modelId);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        if (termsQuery != null) {
+            return this;
+        }
+
+        if (weightedTokensSupplier != null) {
```

Comment:
Here, you should also check if `weightedTokensSupplier.get() == null` as the async rewrite might have happened, but the fetch not complete.
```suggestion
        if (weightedTokensSupplier != null) {
            if (weightedTokensSupplier.get() == null) { 
                return this; 
            }
```